python/cudf/cudf/core/column/column.py

```diff
@@ -2318,285 +2333,69 @@ def as_column(
-                    and pyarrow_array.null_count
-                ):
-                    pyarrow_array = pyarrow_array.cast("float64").fill_null(
-                        np.nan
+                    pd_dtype = dtype.to_pandas()
+                arbitrary = pd.Series(arbitrary, dtype=pd_dtype)
+            else:
+                if isinstance(dtype, np.dtype):
+                    typ = np_to_pa_dtype(dtype)
+                else:
+                    typ = dtype.to_arrow()
+                try:
+                    arbitrary = pa.array(
+                        arbitrary, type=typ, from_pandas=from_pandas
                     )
-
-                data = as_column(
-                    pyarrow_array,
-                    dtype=dtype,
-                    nan_as_null=nan_as_null,
-                )
-            except (pa.ArrowInvalid, pa.ArrowTypeError, TypeError) as e:
-                if isinstance(e, MixedTypeError):
-                    raise TypeError(str(e))
-                if is_categorical_dtype(dtype):
-                    sr = pd.Series(arbitrary, dtype="category")
-                    data = as_column(sr, nan_as_null=nan_as_null, dtype=dtype)
-                elif np_type == np.str_:
-                    sr = pd.Series(arbitrary, dtype="str")
-                    data = as_column(sr, nan_as_null=nan_as_null)
-                elif is_interval_dtype(dtype):
-                    sr = pd.Series(arbitrary, dtype="interval")
-                    data = as_column(sr, nan_as_null=nan_as_null, dtype=dtype)
-                elif (
-                    isinstance(arbitrary, Sequence)
-                    and len(arbitrary) > 0
-                    and any(
-                        cudf.utils.dtypes.is_column_like(arb)
-                        for arb in arbitrary
+                except (pa.ArrowInvalid, pa.ArrowTypeError):
+                    if not isinstance(dtype, np.dtype):
+                        dtype = dtype.to_pandas()
+                    arbitrary = pd.Series(arbitrary, dtype=dtype)
+            data = as_column(arbitrary, nan_as_null=nan_as_null)
+        else:
+            try:
+                arbitrary = pa.array(arbitrary, from_pandas=from_pandas)
+                if cudf.get_option(
+                    "default_integer_bitwidth"
+                ) and pa.types.is_integer(arbitrary.type):
+                    typ = np_to_pa_dtype(_maybe_convert_to_default_type("int"))
+                    arbitrary = arbitrary.cast(typ)
+                elif cudf.get_option(
+                    "default_float_bitwidth"
+                ) and pa.types.is_floating(arbitrary.type):
+                    typ = np_to_pa_dtype(
+                        _maybe_convert_to_default_type("float")
                     )
-                ):
+                    arbitrary = arbitrary.cast(typ)
+            except (pa.ArrowInvalid, pa.ArrowTypeError):
+                if any(is_column_like(val) for val in arbitrary):
```

Comment:
I believe so since an initial `pa.array` call proxy-checks that, but this would fail here for generators for example so I'll make sure this branch casts the input to list or similar